### ISSUE TITLE
Automatically respond to crates.io API token leaks on GitHub (via GitHub secret scanning)

### ISSUE THREAD

**Issue by greysteil** [rust-lang/crates.io#3400]
**Is your feature request related to a problem? Please describe.**
Token leaks [happen](https://github.com/ChALkeR/notes/blob/master/Do-not-underestimate-credentials-leaks.md). There's been previous discussion about [avoiding them in cargo](https://github.com/rust-lang/cargo/issues/2063) but I couldn't find any discussion about automatically responding to them.

I'd like crates.io to automatically revoke any credentials that are leaked into a GitHub public repo. I run the [GitHub secret scanning](https://docs.github.com/en/developers/overview/secret-scanning) team, so I can provide you with a feed of potential matches we find in public repos, if you're willing to take action on them automatically.

We have a similar setup with folks like AWS - if you leak your AWS credentials on GitHub you'll quickly get an email from AWS saying the credential has been quarantined. Under the hood, they're receiving the feed of matches from GitHub.

**Describe the solution you'd like**
crates.io would have an endpoint for receiving details of leaked secrets from GitHub, following the [specification here](https://docs.github.com/en/developers/overview/secret-scanning#create-a-secret-alert-service). crates.io would automatically revoke leaked secrets, and would email their owners to notify them that they had been revoked. The email would include details of where the secret had been exposed (as provided by GitHub).

**Describe alternatives you've considered**
crates.io could notify users of leaked secrets, but not automatically revoke them. However, this would leave leaked credentials exposed.

**Comment by Turbo87** [comment]
@greysteil would https://github.com/rust-lang/crates.io/issues/2653 fit what you're requesting here?

**Comment by greysteil** [comment]
If API tokens were able to self-revoke then we could hit the self-revoke endpoint with our findings, yep. It might feel a bit weird, though - since the tokens we find don't "belong" to us, it's not totally natural that we would presume we can automatically revoke them. They do "belong" to crates (together with the user), so it feels more natural that you would initiate the revoke (and GitHub is just informing you of the leak).

**Comment by Turbo87** [comment]
@greysteil we've discussed this topic in our team meeting today. we generally agree that participating in the automated token scanning would be beneficial. we are however currently in the transition phase from Mozilla to the Rust Foundation (see https://foundation.rust-lang.org), and there are a few open question that the foundation would have to answer before we can move forward here.

in other words: we will have to put this issue on hold for a few weeks until the situation is a bit more structured, but we do plan to return to this :)

**Comment by greysteil** [comment]
Thanks for the update @Turbo87. Just let me know when you've had a chance do discuss further. :octocat:

**Comment by greysteil** [comment]
Hey team, revisiting this now the Rust Foundation has had a chance to get established :-)

**Comment by Turbo87** [comment]
@greysteil we now have an experimental endpoint deployed to https://crates.io/api/github/secret-scanning/verify. do you have a way to test this on your end? can we assist with that in some way?

**Comment by greysteil** [comment]
That's great! OK, on the GitHub side we just need to add your endpoint to our detection service and deploy that change. I'll get it done today. Once that's done you'll be able to test in production by committing a token (for an account which doesn't have any associated crates).

I'll let you know when everything is deployed.

**Comment by Turbo87** [comment]
not sure if you already know this, but our current token format is `cio[a-zA-Z0-9]{32}` :)

**Comment by greysteil** [comment]
Yep! And all deployed - you should be able to test now.

A couple of follow ups:
* There are some (very straightforward) legals associated with the programme that basically say GitHub is allowed to use any information you provide us (like the regex above, or details of whether a detection is a true or false positive). Can you email secret-scanning@github.com so we can send those over to you?
* Having scanned for the pattern above for the last few months, we're seeing about 20% of findings are true positives. Any chance of updating the pattern to make it more identifiable (e.g., with an `_` following the prefix and perhaps a checksum)? If so we'd be able to add this pattern to push protection (which we'd like to roll out to public repos in 2023)

**Comment by greysteil** [comment]
Also, I just got the following email, so I'm guessing things are starting to work?! But I didn't just leak a credential (I may have leaked one ages ago, but not sure we would have immediately sent that over to you)...

<img width="424" alt="image" src="https://user-images.githubusercontent.com/1144873/207389757-48127191-633e-4d95-a14e-ae055fd714f1.png">

**Comment by Turbo87** [comment]
```
Dec 13 16:49:28 [app/web.4](https://my.papertrailapp.com/events?q=program%3Aapp%2Fweb.4&focus=1539363865503342657&selected=1539363865503342657)  WARN cargo_registry::controllers::github::secret_scanning: Revoked API token '<redacted>' for user greysteil (19108)
```

another request, apparently with the same token. and it looks like we might not be deduplicating correctly... 😅

**Comment by Turbo87** [comment]
hmm... but then again, what are we supposed to return for tokens that have already been revoked before? false positive or true positive?

I guess we definitely should not send a second email though 🙈

**Comment by greysteil** [comment]
Yeah, I got a second email, this time with a correct link 🥳 

I'd recommend _not_ sending emails for tokens that have already been revoked, though, assuming they can't be unrevoked.

**Comment by marzvrover** [comment]
> Also, I just got the following email, so I'm guessing things are starting to work or you're testing? I didn't just leak a credential (I may have leaked one ages ago, but not sure we would have been immediately sent that over to you)...
> 
> <img alt="image" width="424" src="https://user-images.githubusercontent.com/1144873/207389757-48127191-633e-4d95-a14e-ae055fd714f1.png">

the incorrect link is my fault, I did a test to make sure there was no errors 😅

**Comment by Turbo87** [comment]
> Yeah, I got a second email, this time with a correct link 🥳

ahh, now that makes sense. by posting the log message and forgetting to redact the token initially I triggered your detection again which then caused a request to our system.

https://github.com/rust-lang/crates.io/pull/5641 is changing the logging to no longer include the plaintext token, and https://github.com/rust-lang/crates.io/pull/5642 will change the implementation to no longer send emails for already revoked tokens.

**Comment by marzvrover** [comment]
> I'm guessing that's where you picked up an example token to do your test with?

exactly :D

**Comment by greysteil** [comment]
Awesome! The final step on the GitHub side is for us put a post on the GitHub changelog. Once legals are done one of the team on our side will draft something (here's [the text we used with PyPI](https://github.blog/changelog/2021-03-22-the-python-package-index-is-now-a-github-secret-scanning-integrator/)) and clear it with you before anything is posted.

Thanks for all your work on this @Turbo87! The team on my side are all psyched to help keep Rust users more secure. 🦀

**Comment by Turbo87** [comment]
thank @dangardner! I just pushed it over the finish line 😅 

looks like https://github.com/aashah just ran two more tests:

```
Dec 13 17:46:35 WARN cargo_registry::controllers::github::secret_scanning: Revoked API token gh_login=aashah user_id=157042 token_id=75826
Dec 13 17:46:35 WARN cargo_registry::controllers::github::secret_scanning: Failed to send email notification gh_login=aashah user_id=157042 error="No address found"
Dec 13 17:47:49 WARN cargo_registry::controllers::github::secret_scanning: Revoked API token gh_login=aashah user_id=157042 token_id=75827
```

**Comment by Turbo87** [comment]
> Having scanned for the pattern above for the last few months, we're seeing about 20% of findings are true positives. 

I'm only seeing four true positive tokens so far. how did you get to the 20% number?

> Any chance of updating the pattern to make it more identifiable (e.g., with an `_` following the prefix and perhaps a checksum)? If so we'd be able to add this pattern to push protection (which we'd like to roll out to public repos in 2023)

unless I'm missing something, this would mean invalidating all existing tokens. we could potentially consider it for new tokens, but then the existing old tokens would fall out of that protection. not sure if there is a good way forward here 🤔

**Comment by pietroalbini** [comment]
Drive-by comment, if we were to change the token format it should probably be `cratesio_token_RANDOM`, so that new token formats could be easily added (for example `cratesio_foo_RANDOM`).

**Comment by Turbo87** [comment]
@marzvrover 👋 we noticed last week that some of the API requests from GitHub to crates.io have been failing because the requests were using the generic user agent header: `Go-http-client/2.0`. we regularly block traffic with generic user agents and it took us quite a while to notice that this also blocked the secret scanning requests. would it be viable on your side to use a more descriptive user agent for these kinds of requests? :)
